week_2/dagster_ucr/project/week_2.py

```diff
@@ -3,29 +3,55 @@
 from dagster import In, Nothing, Out, ResourceDefinition, graph, op
 from dagster_ucr.project.types import Aggregation, Stock
 from dagster_ucr.resources import mock_s3_resource, redis_resource, s3_resource
+from dagster import get_dagster_logger
+from datetime import datetime as dt
 
+logger = get_dagster_logger()
 
-@op
-def get_s3_data():
-    pass
 
+@op(required_resource_keys={"s3"},
+    config_schema={"s3_key": str},
+    out={"stocks": Out(dagster_type=List[Stock])},
+    tags={"kind": "s3"},
+    description="Get a list of stocks from an S3 file",
+)
+def get_s3_data(context):
+    output = list()
+    data = context.resources.s3.get_data(context.op_config["s3_key"])
+    for row in data:
+        stock = Stock.from_list(row)
+        output.append(stock)
+    return output
 
-@op
-def process_data():
-    # Use your op from week 1
-    pass
 
+@op(
+    ins={"stocks": In(dagster_type=List[Stock])},
+    out={"aggregation": Out(dagster_type=Aggregation)},
+    description="Aggregate stock data",
+)
+def process_data(stocks: List[Stock]) -> Aggregation:
+    top_high = sorted(stocks, key=lambda x: x.high, reverse=True)[0]
```

Comment:
not critical, but in order to find the max you don't have to sort (~O(NLOGN)). Instead you can iterate over the array and keep track of the highest number. I believe that's what the `max` method does. For example:
```
ighest_stock: Stock = max(stocks, key=attrgetter("high"))
```

instead of attrgetter you can use lamba x: x.high

Reply:
@OrenLederman Thanks! How were you able to use max method to get top N records for week 1 challenge?